tests/drivers/counter/counter_basic_api/src/test_counter.c

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <counter.h>
+#include <ztest.h>
+#include <kernel.h>
+
+static volatile u32_t top_cnt;
+static volatile u32_t alarm_cnt;
+
+static void top_handler(struct device *dev, void *user_data);
+
+void *exp_user_data = (void *)199;
+
+#ifdef CONFIG_COUNTER_MCUX_RTC
+#define COUNTER_PERIOD_US (2*USEC_PER_SEC)
+#else
+#define COUNTER_PERIOD_US 20000
+#endif
+
+struct counter_alarm_cfg alarm_cfg;
+struct counter_alarm_cfg alarm_cfg2;
+
+const char *devices[] = {
+
+#ifdef CONFIG_COUNTER_TIMER0
+	/* Nordic TIMER0 may be reserved for Bluetooth */
+	CONFIG_COUNTER_TIMER0_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER1
+	CONFIG_COUNTER_TIMER1_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER2
+	CONFIG_COUNTER_TIMER2_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER3
+	CONFIG_COUNTER_TIMER3_NAME,
+#endif
+#ifdef CONFIG_COUNTER_TIMER4
+	CONFIG_COUNTER_TIMER4_NAME,
+#endif
+#ifdef CONFIG_COUNTER_RTC0
+	/* Nordic RTC0 may be reserved for Bluetooth */
+	CONFIG_COUNTER_RTC0_NAME,
+#endif
+	/* Nordic RTC1 is used for the system clock */
+#ifdef CONFIG_COUNTER_RTC2
+	CONFIG_COUNTER_RTC2_NAME,
+#endif
+#ifdef CONFIG_COUNTER_IMX_EPIT_1
+	EPIT_1_LABEL,
+#endif
+#ifdef CONFIG_COUNTER_IMX_EPIT_2
+	EPIT_2_LABEL,
+#endif
+#ifdef DT_RTC_MCUX_0_NAME
+	DT_RTC_MCUX_0_NAME,
+#endif
+};
+typedef void (*counter_test_func_t)(const char *dev_name);
+
+
+static void counter_tear_down_instance(const char *dev_name)
+{
+	int err;
+	struct device *dev;
+
+	dev = device_get_binding(dev_name);
+
+	err = counter_set_top_value(dev, counter_get_max_top_value(dev),
+				    NULL, NULL);
+	zassert_equal(0, err, "Setting top value to default failed\n");
+
+	err = counter_stop(dev);
+	zassert_equal(0, err, "Counter failed to stop\n");
+}
+
+static void test_all_instances(counter_test_func_t func)
+{
+	for (int i = 0; i < ARRAY_SIZE(devices); i++) {
```

Comment:
If ARRAY_SIZE() is 0 we should report an error, since no tests are run.  Right now it looks like things pass.